jax/_src/prng.py

```diff
@@ -0,0 +1,475 @@
+    # instead of a jnp.ndarray due to tree_unflatten
+    if (type(key_data) not in [object, bool] and
+        not _is_prng_key_data(impl, key_data)):
+      raise TypeError(
+          f'Invalid PRNG key data {key_data} for PRNG implementation {impl}')
+    self.impl = impl
+    self.keys = key_data
+
+  def tree_flatten(self):
+    return (self.keys,), self.impl
+
+  @classmethod
+  def tree_unflatten(cls, impl, keys):
+    keys, = keys
+    return cls(impl, keys)
+
+  @property
+  def dtype(self):
+    # TODO(frostig): remove after deprecation window
+    if config.jax_enable_custom_prng:
+      raise AttributeError("'PRNGKeyArray' has no attribute 'dtype'")
+    else:
+      warnings.warn(
+          'deprecated `dtype` attribute of PRNG key arrays', FutureWarning)
+      return np.uint32
+
+  def shape(self):
+    # TODO(frostig): simplify once we always enable_custom_prng
+    if config.jax_enable_custom_prng:
+      return self._shape
+    else:
+      warnings.warn(
+          'deprecated `shape` attribute of PRNG key arrays. In a future version '
+          'of JAX this attribute will be removed or its value may change.',
+          FutureWarning)
+      return self.keys.shape
+
+  @property
+  def _shape(self):
+    base_ndim = len(self.impl.key_shape)
+    return self.keys.shape[:-base_ndim]
+
+  def _is_scalar(self):
+    base_ndim = len(self.impl.key_shape)
+    return self.keys.ndim == base_ndim
+
+  def __len__(self):
+    if self._is_scalar():
+      raise TypeError('len() of unsized object')
+    return len(self.keys)
+
+  def __iter__(self) -> Iterator['PRNGKeyArray']:
+    if self._is_scalar():
+      raise TypeError('iteration over a 0-d single PRNG key')
+    return (PRNGKeyArray(self.impl, k) for k in iter(self.keys))
+
+  def __getitem__(self, idx) -> 'PRNGKeyArray':
+    if not isinstance(idx, tuple):
+      idx = (idx,)
+    if any(type(i) is not int for i in idx):
+      raise NotImplementedError(
+          'PRNGKeyArray only supports indexing with integer indices. '
+          f'Cannot index at {idx}')
+    base_ndim = len(self.impl.key_shape)
+    ndim = self.keys.ndim - base_ndim
+    if len(idx) > ndim:
+      raise IndexError(
+          f'too many indices for PRNGKeyArray: array is {ndim}-dimensional '
+          f'but {len(idx)} were indexed')
+    return PRNGKeyArray(self.impl, self.keys[idx])
+
+  def _fold_in(self, data: int) -> 'PRNGKeyArray':
+    return PRNGKeyArray(self.impl, self.impl.fold_in(self.keys, data))
+
+  def _random_bits(self, bit_width, shape) -> jnp.ndarray:
+    return self.impl.random_bits(self.keys, bit_width, shape)
+
+  def _split(self, num: int) -> 'PRNGKeyArray':
+    return PRNGKeyArray(self.impl, self.impl.split(self.keys, num))
+
+  def __repr__(self):
+    arr_shape = self._shape
+    pp_keys = pp('shape = ') >> pp(arr_shape)
+    if isinstance(self.impl, PRNGImpl):
+      pp_impl = self.impl.pprint()
+    else:
+      pp_impl = pp(self.impl)
```

Comment:
Isn't `self.impl` always a `PRNGImpl`?

Reply:
This was intended to allow for a PRNG implementation that duck-types as a `PRNGImpl` but need not have a `pprint` , but on second thought that's unnecessary complexity without a known need. Good catch. Removed.